### ISSUE TITLE
Migrate to pkg:web

### ISSUE THREAD

**Issue by kevmoo** [flutter/devtools#6606]
Required to support compilation to wasm

- [x] Migrate pkg:devtools_app_shared - WIP - https://github.com/flutter/devtools/pull/6626
- [x] Migrate pkg:devtools_app

Googlers: [go/wasm-ready-pkg-migrate](https://goto.google.com/wasm-ready-pkg-migrate)

**Comment by kenzieschmoll** [comment]
migrations for `devtools_app` that are not straightforward, or just need a little more attention to figure out:

- [ ] _drag_and_drop_web.dart: uses `MouseEvent.offset`
- [ ] _framework_initialize_web.dart: `for (var element in html.document.body!.querySelectorAll('.legacy-dart'))`
- [ ] _export_web.dart: uses `Url.createObjectUrl`
- [ ] allowed_error_html.dart: uses `window.console`
- [ ] _logger_html.dart: uses `window.console`
- [ ] _server_web.dart: there is an outstanding todo to use package:http instead of dart:html here for http requests
- [ ] post_message_web.dart: uses `html.window.onMessage.map(...)`

**Comment by kevmoo** [comment]
Console:

https://pub.dev/documentation/web/0.3.0/helpers/console.html
https://pub.dev/documentation/web/0.3.0/helpers/$ConsoleExtension.html

**Comment by srujzs** [comment]
- `MouseEvent.offset`: The `dart:html` version returns a `Point` with `offsetX` and `offsetY`. There's some checking to make sure `offsetX` is supported, but that's stale since modern browsers should support it: https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/offsetX. If the `Point` class is heavily used in devtools, then maybe we should add it as a helper class in `package:web`. If not, using a record with `offsetX` and `offsetY` should probably suffice.
- `createObjectURL`: The `dart:html` version checks for a `webkitURL`. This is no longer needed as it seems all browsers nowadays support `URL`: https://developer.mozilla.org/en-US/docs/Web/API/URL so it should be safe to url the `package:web` version => `URL.createObjectURL`.
- `html.window.onMessage`: I think this needs `_ElementEventStreamImpl` to be added to the helpers in `package:web`. A workaround is using event listeners, but `Stream`s are nicer.

**Comment by kenzieschmoll** [comment]
@srujzs 
re: MouseEvent, what about the `dataTransfer` property?
`event.dataTransfer.dropEffect = 'move';`

**Comment by srujzs** [comment]
- I think you'll need to downcast `MouseEvent` to `DragEvent` to get `dataTransfer`, which then has a `dropEffect` property.

- I think you want to avoid importing `dart:js` here. The type you want is `JSArray` (I know the naming is unfortunate, but alas). To convert from a list to one, you need a `List<JSAny>`. This can be created by making `content` itself a `JSString`: `[content.toJS].toJS` should get you want you the `JSArray` you need.

**Comment by kenzieschmoll** [comment]
Thanks @srujzs. Getting closer! A few more errors to go:
- `reader` is a `FileReader` object, and `droppedFile` is a `File` object:
![Screenshot 2023-11-09 at 11 29 03 PM](https://github.com/flutter/devtools/assets/43759233/1be52460-31ad-4313-b46a-f48efd86ac70)
- `element.style.display = 'none'`
- `document.body!.append(element);`
- `element.click()`
![Screenshot 2023-11-09 at 11 31 26 PM](https://github.com/flutter/devtools/assets/43759233/1a27be33-7cd1-45a6-b279-2afd67554aed)

**Comment by srujzs** [comment]
Nice!

- I think this is similar to the other issues where this stream getter is not available (`onLoadEnd`). It does seems to exist on `Element`, so a hacky workaround until we publish a version with it is to just cast to `Element` and use `onLoad` (note that we don't check whether that cast is correct since they're both just some random JS objects in the type system's eyes).
- `lastModifiedData` is deprecated in the browser. The right equivalent here is using `lastModified` to get the ms since epoch to construct a `DateTime` e.g. `DateTime.fromMillisecondsSinceEpoch(droppedFile.lastModifiedDate, isUtc: true)`.
- You'll likely need to downcast here to the `Element` type you want in order to use style. I believe downcasting `element` to `HTMLAnchorElement` should resolve the rest. Using `createElement` and then downcasting is a downgrade from the nice `dart:html` constructors, and maybe we should have a `createElementAs<HTMLAnchorElement>()`-like helper somewhere.

**Comment by kevmoo** [comment]
Closing this out as completed via https://github.com/flutter/devtools/pull/6626 - still waiting for dependencies to migrate so we can build w/ wasm

**Comment by kenzieschmoll** [comment]
Everything has been migrated to package:web except for one file: https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/shared/config_specific/post_message/_post_message_web.dart#L5-L9. Removing this is blocked on a new version of `package:web` getting rolled into flutter.

However, if we want to start experimenting with dart2wasm now, apply the changes from this patch and it should compile fine: https://github.com/flutter/devtools/compare/master...kenzieschmoll:dart2wasm-experiment?expand=1. 

@kevmoo

**Comment by kevmoo** [comment]
Reopening then. To stay on top of https://github.com/flutter/devtools/issues/6606#issuecomment-1809080582

**Comment by kevmoo** [comment]
Glad we're getting numbers on this. I think the only concrete thing blocking us is either migrating the inline `gtags` or moving to `unified_analytics`.

Is the latter tracked anywhere?

**Comment by kenzieschmoll** [comment]
migrating to unified_analytics is tracked here: https://github.com/flutter/devtools/issues/6242. This is something that @eliasyishak may be picking up in the short term.
